### PR TITLE
[SPARK-24718][SQL] Timestamp support pushdown to parquet data source

### PR DESCRIPTION
**Pull request by wangyum**
## What changes were proposed in this pull request?

`Timestamp` support pushdown to parquet data source.
Only `TIMESTAMP_MICROS` and `TIMESTAMP_MILLIS` support push down.

## How was this patch tested?

unit tests and benchmark tests

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -378,6 +378,15 @@ object SQLConf {
...
     .createWithDefault(true)
 
+  val PARQUET_FILTER_PUSHDOWN_TIMESTAMP_ENABLED =
+    buildConf("spark.sql.parquet.filterPushdown.timestamp")
+      .doc("If true, enables Parquet filter push-down optimization for Timestamp. " +
+        "This configuration only has an effect when 'spark.sql.parquet.filterPushdown' is " +
+        "enabled and Timestamp stored as TIMESTAMP_MICROS or TIMESTAMP_MILLIS type.")
+    .internal()
+    .booleanConf
+    .createWithDefault(true)
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: May be default should be `false`. because `PARQUET_OUTPUT_TIMESTAMP_TYPE` default is `INT96`.

**Comment by wangyum** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala [hunk 2]
```diff
@@ -387,6 +389,82 @@ class ParquetFilterSuite extends QueryTest with ParquetTest with SharedSQLContex
...
 
+  test("filter pushdown - timestamp(TIMESTAMP_MILLIS)") {
+    val ts1 = Timestamp.valueOf("2018-06-14 08:28:53.123")
+    val ts2 = Timestamp.valueOf("2018-06-15 08:28:53.123")
+    val ts3 = Timestamp.valueOf("2018-06-16 08:28:53.123")
+    val ts4 = Timestamp.valueOf("2018-06-17 08:28:53.123")
+
+    val data = Seq(ts1, ts2, ts3, ts4)
+
+    withSQLConf(SQLConf.PARQUET_OUTPUT_TIMESTAMP_TYPE.key ->
```
Comment (gengliangwang) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala]: This case is quite similar to the one below. Should we use a loop for setting the key `SQLConf.PARQUET_OUTPUT_TIMESTAMP_TYPE.key` to avoid duplicated code.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala [hunk 3]
```diff
@@ -387,6 +389,82 @@ class ParquetFilterSuite extends QueryTest with ParquetTest with SharedSQLContex
     }
   }
 
+  test("filter pushdown - timestamp(TIMESTAMP_MILLIS)") {
```
Comment (gengliangwang) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala]: I think we should also test INT96 timestamp type. Also maybe when `PARQUET_FILTER_PUSHDOWN_TIMESTAMP_ENABLED` is disabled.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala (hunk 2 above)
Comment (wangyum) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala]: I changed to:
```scala
    // spark.sql.parquet.outputTimestampType = TIMESTAMP_MILLIS
    val millisData = Seq(Timestamp.valueOf("2018-06-14 08:28:53.123"),
      Timestamp.valueOf("2018-06-15 08:28:53.123"),
      Timestamp.valueOf("2018-06-16 08:28:53.123"),
      Timestamp.valueOf("2018-06-17 08:28:53.123"))
    withSQLConf(SQLConf.PARQUET_OUTPUT_TIMESTAMP_TYPE.key ->
      ParquetOutputTimestampType.TIMESTAMP_MILLIS.toString) {
      testTimestampPushdown(millisData)
    }

    // spark.sql.parquet.outputTimestampType = TIMESTAMP_MICROS
    val microsData = Seq(Timestamp.valueOf("2018-06-14 08:28:53.123456"),
      Timestamp.valueOf("2018-06-15 08:28:53.123456"),
      Timestamp.valueOf("2018-06-16 08:28:53.123456"),
      Timestamp.valueOf("2018-06-17 08:28:53.123456"))
    withSQLConf(SQLConf.PARQUET_OUTPUT_TIMESTAMP_TYPE.key ->
      ParquetOutputTimestampType.TIMESTAMP_MICROS.toString) {
      testTimestampPushdown(microsData)
    }
```
We shouldn't use same data to test `TIMESTAMP_MILLIS` type and `TIMESTAMP_MICROS` type:
1.  `TIMESTAMP_MILLIS` type will truncate `456` if use  `microsData` to test.
2. It can't test `DateTimeUtils.fromJavaTimestamp(t.asInstanceOf[Timestamp]` if use `millisData`.

**Review by gengliangwang — APPROVED** [review]
LGTM

**Comment by gatorsmile** [comment]
cc @michal-databricks @mswit-databricks @rdblue @cloud-fan

File: sql/core/benchmarks/FilterPushdownBenchmark-results.txt [hunk 4]
```diff
@@ -578,3 +578,127 @@ Native ORC Vectorized                       11622 / 12196          1.4         7
...
 
 
+================================================================================================
+Pushdown benchmark for Timestamp
+================================================================================================
+
+Java HotSpot(TM) 64-Bit Server VM 1.8.0_151-b12 on Mac OS X 10.12.6
+Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz
+
+Select 1 timestamp stored as INT96 row (value = CAST(7864320 AS timestamp)): Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
```
Comment (cloud-fan) [inline sql/core/benchmarks/FilterPushdownBenchmark-results.txt]: shall we add a new line after the benchmark name? e.g.
```
Select 1 timestamp stored as INT96 row (value = CAST(7864320 AS timestamp)):
Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
...
```

We can send a follow-up PR to fix this entire file.

**Comment by cloud-fan** [comment]
LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 5]
```diff
@@ -378,6 +378,15 @@ object SQLConf {
     .booleanConf
     .createWithDefault(true)
 
+  val PARQUET_FILTER_PUSHDOWN_TIMESTAMP_ENABLED =
+    buildConf("spark.sql.parquet.filterPushdown.timestamp")
+      .doc("If true, enables Parquet filter push-down optimization for Timestamp. " +
+        "This configuration only has an effect when 'spark.sql.parquet.filterPushdown' is " +
+        "enabled and Timestamp stored as TIMESTAMP_MICROS or TIMESTAMP_MILLIS type.")
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Shell we note `INT64` here?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala [hunk 6]
```diff
@@ -517,7 +585,6 @@ class ParquetFilterSuite extends QueryTest with ParquetTest with SharedSQLContex
     }
   }
 
-
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala]: nit: I would revert this change if you are going to push more changes.

**Review by HyukjinKwon — APPROVED** [review]
LGTM too

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 5 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: I think end users have a better understanding of `TIMESTAMP_MICROS` and `TIMESTAMP_MILLIS`.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala (hunk 6 above)
Comment (wangyum) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala]: OK

File: sql/core/benchmarks/FilterPushdownBenchmark-results.txt (hunk 4 above)
Comment (wangyum) [reply sql/core/benchmarks/FilterPushdownBenchmark-results.txt]: OK. I'll send a follow-up PR.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 5 above)
Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: ... I don't think users will understand any of them ..

Comment (gatorsmile) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: You need to explain how to use `spark.sql.parquet.outputTimestampType` to control the Parquet timestamp type  Spark uses to writes parquet files.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 1 above)
Comment (rdblue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Because we're using the file schema, it doesn't mater what the write configuration is. It only matters what it was when the file was written. If the file has an INT96 timestamp, this should just not push anything down.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 5 above)
Comment (rdblue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: I would just note that push-down doesn't work for INT96 timestamps in the file. It should work for the others.

**Comment by rdblue** [comment]
+1

**Comment by HyukjinKwon** [comment]
Merged to master.
